Code generation with GYB

adamkuipers opened the pull request:
Sorry I left this hanging for so long. I took a break from computers and then forgot about it. Let me know if you guys are still interested and I'll tidy up the commits.

What's in this pull request?
============================

Generates definitions of zip and map for arities up to 22 (configurable in .gyb template with max_arity).

Why merge this pull request?
============================

Allows changing the arity of map and zip functions without duplicating any code. The primitive footprint for the API is also reduced with all zip functions based on zip(_:_:). There are now nine fewer functions that call the low level Gen(unGen:) initializer.

What's worth discussing about this pull request?
================================================

We might want other derived functions like ap generated to max_arity number of definitions. Also, I think I solved the build issues by moving the templates out of Sources. We'll see what travis has to say about this.

What downsides are there to merging this pull request?
======================================================

The downside of generated code, which is that it's mechanized instead of an abstraction through code. It's harder to develop a template than regular code since there's a level of indirection when debugging.

adamkuipers commented:
Looks like all checks pass now.

CodaFi commented on Templates/Cartesian.swift.gyb:
```diff
@@ -0,0 +1,72 @@
+MAX_ARITY = 22
+}%
+
+extension Gen /*: Cartesian*/ {
```
Could you add a comment here or in the file header that this file is machine generated and should not be edited?

adamkuipers commented:
Added comments and some tests. Also, I changed the build to no longer fetch GYB from the Swift repo before each build, instead keeping it in Utilities.

CodaFi commented:
Really just one more minor thing and this is good to go.  I wish there were some way for us to take a dependency on just gyb but that may be a bad idea considering how small the script is.

adamkuipers commented:
Sorry, but what do you mean by "take a dependency on just gyb?"

CodaFi commented:
Like if there were a way to just add a submodule for the worktree for just GYB.  But that's just me thinking out loud.  This is fine.

CodaFi commented:
Thanks @adamkuipers 

⛵
